src/journal/compress.c

```diff
@@ -541,6 +645,90 @@ int compress_stream_lz4(int fdf, int fdt, uint64_t max_bytes) {
 #endif
 }
 
+int compress_stream_zstd(int fdf, int fdt, uint64_t max_bytes) {
+#if HAVE_ZSTD
+        _cleanup_(ZSTD_freeCCtxp) ZSTD_CCtx *s = ZSTD_createCStream();
+        size_t out_size = ZSTD_CStreamOutSize();
+        uint8_t out[out_size];
+        char *src = NULL;
+        struct stat st;
+        ZSTD_inBuffer ZSTD_in = {src, 0, 0};
+        ZSTD_outBuffer ZSTD_out = {&out, sizeof(out), 0};
+        size_t ret;
+        uint64_t left = max_bytes;
+        int r;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        if (!s)
+                return -ENOMEM;
+
+        ret = ZSTD_initCStream(s, ZSTD_COMPRESSION_LEVEL);
+        if (ZSTD_isError(ret)) {
+                log_error("Failed to initialize ZSTD encoder: %s", ZSTD_getErrorName(ret));
+                return zstd_ret_to_errnoh(ret);
+        }
+
+        if (fstat(fdf, &st) < 0)
+                return log_debug_errno(errno, "fstat() failed: %m");
+        ZSTD_in.size = st.st_size;
+
+        src = mmap(NULL, st.st_size, PROT_READ, MAP_PRIVATE, fdf, 0);
+        if (src == MAP_FAILED)
+                return -errno;
+        ZSTD_in.src = src;
+
+        for (;;) {
+                while (ZSTD_out.pos < ZSTD_out.size) {
+                        ret = ZSTD_compressStream(s, &ZSTD_out, &ZSTD_in);
+                        if (ZSTD_isError(ret)) {
+                                log_error("ZSTD encoder failed: %s", ZSTD_getErrorName(ret));
+                                return zstd_ret_to_errnoh(ret);
+                        }
+
+                        if (ZSTD_in.pos == ZSTD_in.size)
+                                break;
+                }
+
+                if (left < ZSTD_out.pos)
+                        return -EFBIG;
+
+                r = loop_write(fdt, ZSTD_out.dst, ZSTD_out.pos, false);
+                if (r < 0)
+                        goto cleanup;
+
+                left -= ZSTD_out.pos;
+                ZSTD_out.pos = 0;
+                if (ZSTD_in.pos == ZSTD_in.size)
+                        break;
+        }
+        ret = ZSTD_endStream(s, &ZSTD_out);
```

Comment:
If you have a smaller buffer size than `ZSTD_CStreamOutSize()`, the first call to `ZSTD_endStream()` might not be able to flush all the data, so you will need to call it in a loop until it returns 0.